File: R/prepare_kegg_nonmodel.R
# ---------------------------------------------------------------------------
# prepare_kegg_nonmodel.R
# ---------------------------------------------------------------------------
# Non-model organism KEGG annotation, driven by a KAAS result file.
# R re-implementation of Retrieve_KEGG_annot_nonmodel_org.pl, with the KEGG
# reference tables downloaded automatically (see download_kegg.R).
#
# Inputs
#   kaas_file : KAAS output, 2 columns  <transcript/gene id>\t<K number>
#               (rows without a K number are ignored)
# Outputs (in out_dir)
#   KEGG_pathway2gene.tab   (v1, index)     -> enrichment
#   KEGG_pathway2name.tab   (pathway, info) -> enrichment
#   KEGG_annot_genes.txt                    -> descriptive annotation
# ---------------------------------------------------------------------------

prepare_kegg_nonmodel <- function(kaas_file,
                                   out_dir       = "results",
                                   cache_dir     = "data",
                                   refresh       = FALSE,
                                   strip_isoform = TRUE,
                                   gene_list     = NULL) {

  stopifnot(file.exists(kaas_file))
  ensure_dir(out_dir)

  ## 1. gene -> {KO} from the KAAS result -----------------------------------
  log_msg("reading KAAS result: ", kaas_file)
  kaas <- utils::read.delim(kaas_file, header = FALSE, sep = "\t",
                            stringsAsFactors = FALSE, quote = "",
                            col.names = c("contig", "ko"), fill = TRUE)
  kaas <- kaas[!is.na(kaas$ko) & nzchar(kaas$ko), , drop = FALSE]
  if (strip_isoform)
    kaas$contig <- sub("_i\\d+$", "", kaas$contig)   # collapse isoforms to gene
  kaas$ko <- sub("^ko:", "", kaas$ko)
  kaas <- unique(kaas)

  gene2ko <- split(kaas$ko, kaas$contig)             # named list gene -> KOs
  log_msg("genes with >=1 KO: ", length(gene2ko))

  ## 2. KEGG reference tables (auto-download + parse) ------------------------
  kegg  <- download_kegg_rest(cache_dir, refresh = refresh)
  ko2info <- parse_ko_to_name(kegg$ko_to_name)
  ko2path <- parse_ko_to_path(kegg$ko_to_path)
  path2nm <- parse_pathway_names(kegg$pathway_names)

  ## 3. enrichment file: pathway -> name ------------------------------------
  write_pathway2name(list(names(path2nm), unname(path2nm)),
                     file.path(out_dir, "KEGG_pathway2name.tab"))

  ## 4. enrichment file: pathway -> gene ------------------------------------
  p2g_path <- character(0)
  p2g_gene <- character(0)
  for (gene in names(gene2ko)) {
    paths <- unique(unlist(ko2path[gene2ko[[gene]]], use.names = FALSE))
    paths <- paths[!is.na(paths)]
    if (length(paths)) {
      p2g_path <- c(p2g_path, paths)
      p2g_gene <- c(p2g_gene, rep(gene, length(paths)))
    }
  }
  write_pathway2gene(list(p2g_path, p2g_gene),
                     file.path(out_dir, "KEGG_pathway2gene.tab"))

  ## 5. descriptive annotation ----------------------------------------------
  genes <- if (is.null(gene_list)) names(gene2ko) else gene_list
  annot <- build_kegg_annot(genes, gene2ko, ko2info, ko2path, path2nm)
  annot_file <- file.path(out_dir, "KEGG_annot_genes.txt")
  write_tab(annot, annot_file)

  invisible(list(
    pathway2gene = file.path(out_dir, "KEGG_pathway2gene.tab"),
    pathway2name = file.path(out_dir, "KEGG_pathway2name.tab"),
    annotation   = annot_file
  ))
}

# per-gene descriptive table (values joined with " | " as in the Perl version)
build_kegg_annot <- function(genes, gene2ko, ko2info, ko2path, path2nm) {
  j <- function(x) paste(x, collapse = " | ")
  rows <- lapply(genes, function(gene) {
    kos <- gene2ko[[gene]]
    if (is.null(kos)) kos <- character(0)
    names  <- vapply(kos, function(k) if (!is.null(ko2info[[k]])) ko2info[[k]]$names else "", "")
    titles <- vapply(kos, function(k) if (!is.null(ko2info[[k]])) ko2info[[k]]$title else "", "")
    ecs    <- vapply(kos, function(k) if (!is.null(ko2info[[k]])) ko2info[[k]]$ec    else "", "")
    paths  <- unique(unlist(ko2path[kos], use.names = FALSE))
    paths  <- paths[!is.na(paths)]
    pnames <- unname(path2nm[paths]); pnames[is.na(pnames)] <- ""
    data.frame(
      Gene              = gene,
      KEGG_ID           = j(kos),
      KEGG_names        = j(names),
      KEGG_description  = j(titles),
      EC_number         = j(ecs),
      Pathway_IDs       = j(paths),
      Pathway_names     = j(pnames),
      stringsAsFactors  = FALSE
    )
  })
  do.call(rbind, rows)
}

File: src/prepare_kegg_nonmodel.py
"""Non-model organism KEGG: KAAS query.ko.txt -> enrichment + annotation files."""

import os
import re
from collections import defaultdict

from .download_kegg import download_kegg_rest, parse_ko_to_name, parse_ko_to_path, parse_pathway_names
from .utils import log_msg, write_gmt, write_pathway2gene, write_pathway2name, write_tab


def prepare_kegg_nonmodel(kaas_file, out_dir, cache_dir, refresh=False,
                          strip_isoform=True, gene_list=None):
    gene2ko = defaultdict(set)
    with open(kaas_file) as f:
        for line in f:
            line = line.rstrip("\n")
            if not line:
                continue
            parts = line.split("\t", 1)
            contig = parts[0]
            ko = parts[1].strip() if len(parts) > 1 else ""
            if not ko:
                continue
            if strip_isoform:
                contig = re.sub(r"_i\d+$", "", contig)
            ko = ko.replace("ko:", "")
            gene2ko[contig].add(ko)

    log_msg("KAAS input: ", len(gene2ko), " genes with KO assignments")

    ko_name_file, ko_path_file, pw_name_file = download_kegg_rest(cache_dir, refresh)
    ko2info = parse_ko_to_name(ko_name_file)
    ko2path = parse_ko_to_path(ko_path_file)
    path_names = parse_pathway_names(pw_name_file)

    pw_name_pairs = [(pid, name) for pid, name in sorted(path_names.items())]
    write_pathway2name(pw_name_pairs,
                       os.path.join(out_dir, "KEGG_pathway2name.tab"))

    path2gene_pairs = []
    for gene, kos in gene2ko.items():
        for ko in kos:
            for pth in ko2path.get(ko, []):
                path2gene_pairs.append((pth, gene))

    write_pathway2gene(path2gene_pairs,
                       os.path.join(out_dir, "KEGG_pathway2gene.tab"))
    log_msg("KEGG pathway2gene: ", len(set(p for p, _ in path2gene_pairs)),
            " pathways, ", len(set(g for _, g in path2gene_pairs)), " genes")

    term2genes = defaultdict(list)
    for pth, gene in path2gene_pairs:
        term2genes[pth].append(gene)
    write_gmt(dict(term2genes), path_names,
              os.path.join(out_dir, "KEGG_pathway.gmt"))

    annot_rows = _build_kegg_annot(gene2ko, ko2info, ko2path, path_names,
                                   gene_list)
    annot_cols = ["Gene", "KEGG_ID", "KEGG_names", "KEGG_description",
                  "EC_number", "Pathway_IDs", "Pathway_names"]
    write_tab(annot_rows, annot_cols,
              os.path.join(out_dir, "KEGG_annot_genes.txt"))


def _build_kegg_annot(gene2ko, ko2info, ko2path, path_names, gene_list=None):
    genes = gene_list if gene_list else sorted(gene2ko.keys())
    rows = []
    for gene in genes:
        kos = sorted(gene2ko.get(gene, set()))
        if not kos:
            rows.append((gene, "", "", "", "", "", ""))
            continue
        all_names = []
        all_titles = []
        all_ec = []
        all_pids = set()
        for ko in kos:
            info = ko2info.get(ko, {})
            all_names.append(info.get("names", ""))
            all_titles.append(info.get("title", ""))
            ec = info.get("ec", "")
            if ec:
                all_ec.append(ec)
            for p in ko2path.get(ko, []):
                all_pids.add(p)
        pid_list = sorted(all_pids)
        pnames = [path_names.get(p, "") for p in pid_list]
        rows.append((
            gene,
            " | ".join(kos),
            " | ".join(all_names),
            " | ".join(all_titles),
            " | ".join(all_ec),
            " | ".join(pid_list),
            " | ".join(pnames),
        ))
    return rows
